src/libcharon/plugins/kernel_super/kernel_super_ipsec.c

```diff
@@ -1195,6 +1268,186 @@ static status_t del_routes_internal(private_kernel_super_ipsec_t* this,
+	const struct tunnel_counter_key *tunnel_key = (const struct tunnel_counter_key*)key;
+
+	return chunk_hash_inc(chunk_from_thing(tunnel_key->terminator_id),
+						  chunk_hash_inc(chunk_from_str(tunnel_key->tenant_id),
+						  chunk_hash_inc(chunk_from_str(tunnel_key->account_id),
+						  chunk_hash(chunk_from_str(tunnel_key->vrf_name)))));
+}
+
+
+static void destroy_tunnel_counter_key(void *value, const void *key) {
+	(void)key;
+	if(! value) {
+		return;
+	}
+
+	struct tunnel_counter_key *tunnel_key = (struct tunnel_counter_key *)value;
+	free(tunnel_key->tenant_id);
+	free(tunnel_key->account_id);
+	free(tunnel_key->vrf_name);
+	free(value);
+}
+
+CALLBACK(write_terminator_tenant_account_vrf_tunnels, void,
+		 void *ctx, const struct system_enterprise_map_iterator* iter) {
+	hashtable_t *counter_map = (hashtable_t *)ctx;
+	const char* const tenant_id = iter->lookup_params->tenant_id;
+	const char* const account_id = iter->lookup_params->account_id;
+	const char* const vrf_name = iter->lookup_params->vrf_name;
+	uint32_t term_id = iter->lookup_params->terminator_id;
+
+	struct tunnel_counter_key key = {.terminator_id = term_id,
+									 .tenant_id = tenant_id,
+									 .account_id = account_id,
+									 .vrf_name = vrf_name,
+									 .count = 0};
+	struct tunnel_counter_key *curr_key = counter_map->get(counter_map, &key);
+	if(NULL != curr_key) {
+		key.count = curr_key->count;
+	}
+
+	if (! oper_db_writer_set_terminator_vrf_tunnel_counter(term_id,
+														   tenant_id, account_id, vrf_name,
+														   key.count)) {
+		LOG(DBG_KNL, SL_LOG_ERR, "failed writing vrf: %s in term %u tunnel counter",
+			vrf_name, curr_key->terminator_id);
+	}
+}
+
+CALLBACK(write_terminator_tunnels_oper, void,
+		private_kernel_super_ipsec_t* this,
+		struct pm_terminator_conn* const terminator_conn) {
+	const struct pm_proto_addr* const lookup_addr = picomsg_terminator_conn_get_lookup_addr(terminator_conn);
+	enumerator_t* enumerator = NULL;
+	ike_sa_t* ike_sa;
+	uint64_t term_tunnels_number = 0;
+	hashtable_t *counter_map = NULL;
+	const uint32_t term_id = picomsg_terminator_conn_get_id(terminator_conn);
+
+	counter_map = hashtable_create(tunnel_counter_key_hash, tunnel_counter_key_compare, 10);
+
+	enumerator = charon->controller->create_ike_sa_enumerator(charon->controller, true);
+	while (enumerator->enumerate(enumerator, &ike_sa)) {
+		struct pm_proto_addr ike_sa_addr;
+		host_addr_to_picomsg_addr(ike_sa->get_my_host(ike_sa), &ike_sa_addr);
+		if (! pm_proto_addr_equals(&ike_sa_addr, lookup_addr)) {
+			continue;
+		}
+
+		const ike_super_peer_info_t* const super_peer_info = ike_sa->get_super_peer_info(ike_sa);
+		if (! super_peer_info) {
+			continue;
+		}
+
+		if (IKE_SUPER_PEER_INFO_TYPE_TERM != super_peer_info->get_type(super_peer_info)) {
+			continue;
+		}
+
+		struct tunnel_counter_key *key;
+		INIT(key,
+			 .terminator_id = term_id,
+			 .tenant_id = strdupnull(super_peer_info->get_tenant_id(super_peer_info)),
```

Comment:
Also can defer the string copying and malloc()-ing `key` to happen only for the first time the key is inserted - Start with `key` being a stack variable and holds pointers directly to the strings in super_peer_info, and only if the key is not present in the hashtable then alloc a new key (with `strdup()`-ing the strings) and add it to the table - This exact behaviour was present in the previous implementation of this code, no need to change it.